Cesium3DTilesSelection/src/TilesetContentManager.cpp

```diff
@@ -0,0 +1,1048 @@
+          content,
+          std::move(result.rasterOverlayDetails),
+          pWorkerRenderResources},
+      std::move(result.contentKind));
+
+  if (result.tileInitializer) {
+    result.tileInitializer(tile);
+  }
+
+  tile.setState(TileLoadState::ContentLoaded);
+}
+
+void TilesetContentManager::updateContentLoadedState(
+    Tile& tile,
+    const TilesetOptions& tilesetOptions) {
+  // initialize this tile content first
+  TileContent& content = tile.getContent();
+  if (content.isExternalContent()) {
+    // if tile is external tileset, then it will be refined no matter what
+    tile.setUnconditionallyRefine();
+  } else if (content.isEmptyContent()) {
+    // There are two possible ways to handle a tile with no content:
+    //
+    // 1. Treat it as a placeholder used for more efficient culling, but
+    //    never render it. Refining to this tile is equivalent to refining
+    //    to its children.
+    // 2. Treat it as an indication that nothing need be rendered in this
+    //    area at this level-of-detail. In other words, "render" it as a
+    //    hole. To have this behavior, the tile should _not_ have content at
+    //    all.
+    //
+    // We distinguish whether the tileset creator wanted (1) or (2) by
+    // comparing this tile's geometricError to the geometricError of its
+    // parent tile. If this tile's error is greater than or equal to its
+    // parent, treat it as (1). If it's less, treat it as (2).
+    //
+    // For a tile with no parent there's no difference between the
+    // behaviors.
+    double myGeometricError = tile.getNonZeroGeometricError();
+    const Tile* pAncestor = tile.getParent();
+    while (pAncestor && pAncestor->getUnconditionallyRefine()) {
+      pAncestor = pAncestor->getParent();
+    }
+
+    double parentGeometricError = pAncestor
+                                      ? pAncestor->getNonZeroGeometricError()
+                                      : myGeometricError * 2.0;
+    if (myGeometricError >= parentGeometricError) {
+      tile.setUnconditionallyRefine();
+    }
+  } else if (content.isRenderContent()) {
+    TileRenderContent* pRenderContent = content.getRenderContent();
+    // add copyright
+    pRenderContent->setCredits(GltfUtilities::parseGltfCopyright(
+        *this->_externals.pCreditSystem,
+        pRenderContent->getModel(),
+        tilesetOptions.showCreditsOnScreen));
+
+    // create render resources in the main thread
+    void* pWorkerRenderResources = pRenderContent->getRenderResources();
+    void* pMainThreadRenderResources =
+        this->_externals.pPrepareRendererResources->prepareInMainThread(
+            tile,
+            pWorkerRenderResources);
+    pRenderContent->setRenderResources(pMainThreadRenderResources);
+  }
+
+  tile.setState(TileLoadState::Done);
+}
+
+void TilesetContentManager::updateDoneState(
+    Tile& tile,
+    const TilesetOptions& tilesetOptions) {
+  if (tile.shouldContentContinueUpdated()) {
```

Comment:
`shouldContentContinueUpdated` becomes false as soon as we're able to create children for this tile (or conclusively determine we don't need to do so). But the logic below checks for placeholder raster overlay tile providers and replaces them with real ones once the raster overlay tile provider is ready to go. There's no guarantee that overlays will be ready _before_ the children. So I think this is a bug. A slow raster overlay tile provider may result in some tiles never showing the overlay content.

Reply:
@kring can you please recheck on this one? Based on our offline discussion, it doesn't seem to be a bug right?

Reply:
Yes, I read it as `if (!tile.shouldContentContinueUpdated()) ...`, the opposite of what it really says. Sorry for the noise! Could be worth a comment to explain why this method should only do something when the tile is done updating, but it makes sense to me now at least.